crates/tree-sitter-essence/grammar.js

```diff
@@ -0,0 +1,598 @@
+module.exports = grammar ({
+  name: 'essence',
+
+  extras: $ => [
+    $.single_line_comment,
+    /\s/
+  ],
+
+  rules: {
+    program: $ => repeat(choice(
+      $.find_statement_list,
+      $.constraint_list,
+    )),
+
+    single_line_comment: $ => token(seq(
+      '$',
+      /.*/
+    )),
+
+    find_statement_list: $ => seq(
+      "find",
+      repeat($.find_statement)
+    ),
+
+    find_statement: $ => seq(
+      field("variable_list", $.variable_list),
+      ":",
+      field("domain", $.domain),
+      optional(",")
+    ),
+
+    variable_list: $ => seq(
+      $.variable,
+      optional(repeat(seq(
+        ",",
+        $.variable
+      )))
+    ),
+
+    constraint_list: $ => seq(
+      "such that",
+      repeat($.constraint)
+    ),
+
+    constraint: $ => seq(
+      optional($.not),
+      $.expression,
+      optional(",")
+    ),
+
+    not: $ => "!",
+
+    domain: $ => choice(
+      $.int_domain,
+      $.bool_domain
+    ),
+
+    int_domain: $ => prec.left(seq(
+      "int",
+      optional(seq(
+        "(",
+        $.range_list,
+        //eventually add in expressions here
+        ")"
+      ))
+    )),
+
+    bool_domain: $ => "bool",
+
+    range_list: $ => prec(2, seq(
+      choice(
+      $.lower_bound_range,
+      $.upper_bound_range,
+      $.closed_range,
+      $.integer
+      ),
+      optional(repeat(seq(
+        ",",
+        choice(
+          $.lower_bound_range,
+          $.upper_bound_range,
+          $.closed_range,
+          $.integer
+        )
+      )))
+    )),
+    
+    //should these be expresions not integers?
+    lower_bound_range: $ => seq(
+      $.integer,
+      ".."
+    ),
+
+    upper_bound_range: $ => seq(
+      "..",
+      $.integer
+    ),
+    
+    closed_range: $ => seq(
+      $.integer,
+      "..",
+      $.integer
+    ),
+
+    variable: $ => $.identifier,
+
+    identifier: $ => /[a-zA-Z_][a-zA-Z0-9_]*/,
+
+    expression: $ => seq(
```

Comment:
this conjuction>comparison>addition>term>etc hierarcy seems very low level to me. I checked an tree-sitter seems to support "expression parsing" natively. something like (generic example):

```
    // Binary expressions with operator precedence and associativity
    binary_expression: $ => prec.left(choice(
      prec(1, seq($.expression, '+', $.expression)),   // Addition (left-associative)
      prec(1, seq($.expression, '-', $.expression)),   // Subtraction (left-associative)
      prec(2, seq($.expression, '*', $.expression)),   // Multiplication (left-associative)
      prec(2, seq($.expression, '/', $.expression)),   // Division (left-associative)
      prec.right(3, seq($.expression, '^', $.expression)) // Exponentiation (right-associative)
    )),
```

Essence's operators, their precedences and associativity info can be found in Conjure's code here https://github.com/conjure-cp/conjure/blob/d0721c192c1d63bb65b4aa8131d8fde4b101a3c9/src/Conjure/Language/Expression/Op/Internal/Common.hs#L198-L242

Something like this should make your life easier.

Reply:
I definitely agree. I've been meaning to change it but I was too focused on getting the tests passing. I just fixed it and now it makes more sense and is more intuitive. This example was very helpful, thank you!